### PR TITLE
WIP: Implement a Natural Questions Style Question Answering System

### PR DESCRIPTION
**Pull request by brandenchan**
This implements a Natural Questions style QA system which can return [span, is_impossible, yes, no] predictions. This is done with two prediction heads (QA and text classification) whose outputs are merged using QuestionAnsweringHead.merge().

It is worth mentioning a few limitations in this implementation:

1) Predictions Objects are now implemented for QA but not for any other task
   
2) The structure of Baskets and Samples is different in SQuAD vs NQ and needs to be cleaned in a future PR

3) The apply_tokenization() functions of SQuAD and NQ processors are very similar to each other and might need in future to be merged

4) When performing NQ inference, no probs are returned from the Text Classification Head

5) While data flows end to end, it has not been tested for performance

6) Preprocessing of data is not optimized so NQ is still very slow

### CODE REVIEW

**Comment by brandenchan** [comment]
@tholor Before merging, it would also be good to check that the output of the Squad style Inferencer is still compatible with Haystack. I have a suspicion that the output might be nested in another list() layer

File: farm/modeling/prediction_head.py [hunk 1]
```diff
@@ -1505,3 +1463,42 @@ def label_to_doc_idxs(label, passage_start_t):
 
     def prepare_labels(self, labels, start_of_word, **kwargs):
         return labels
+
+    @staticmethod
+    def merge(preds_all):
```
Comment (tholor) [inline farm/modeling/prediction_head.py:1468]: maybe rename to `merge_formatted_preds()`?

File: farm/data_handler/data_silo.py [hunk 2]
```diff
@@ -163,15 +163,20 @@ def _get_dataset(self, filename, dicts=None):
             datasets = []
 
             desc = f"Preprocessing Dataset"
+            final_tensor_names = None
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/data_handler/data_silo.py]: No longer needed, it's deleted in the latest commit

File: farm/data_handler/input_features.py [hunk 3]
```diff
@@ -449,28 +451,35 @@ def generate_labels(answers, passage_len_t, question_len_t, tokenizer, max_answe
...
 
-
         start_idx = start_vec.index(1)
         end_idx = end_vec.index(1)
 
         label_idxs[i, 0] = start_idx
         label_idxs[i, 1] = end_idx
 
+        # Only Natural Questions trains a classification head on answer_type, SQuAD only has the QA head. answer_type_list
+        # will be None for SQuAD but something like
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/data_handler/input_features.py]: Fixed in latest commit

File: farm/data_handler/processor.py [hunk 4]
```diff
@@ -279,10 +284,6 @@ def _init_samples_in_baskets(self):
             except:
                 logger.error(f"Could not create sample(s) from this dict: \n {basket.raw}")
                 raise
-        baskets_to_remove = [b.id for b in self.baskets if len(b.samples) == 0]
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/data_handler/processor.py:274]: The downsampling fn now makes sure there is at least one sample in the basket

File: farm/data_handler/processor.py [hunk 5]
```diff
@@ -1156,9 +1152,305 @@ def _dict_to_samples(self, dictionary: dict, **kwargs) -> [Sample]:
...
+    def choose_span(sa_text, sa_start_c, la_text, la_start_c):
+        if sa_text:
+            return sa_text, sa_start_c
+        elif la_text:
+            return la_text, la_start_c
+        else:
+            return "", -1
+
+    def unify_short_answers(self, short_answers, doc_text, tok_to_ch):
+        """ In cases where an NQ sample has multiple disjoint short answers, this fn generates the single shortest
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/data_handler/processor.py:1349]: Yes I agree! Will put it on the back log

File: farm/data_handler/samples.py [hunk 6]
```diff
@@ -234,12 +226,85 @@ def process_answers(answers, doc_offsets, passage_start_c, passage_start_t):
...
-                                 "end_t": answer_end_t}
+                                 "end_t": answer_end_t,
+                                 "answer_type": answer["answer_type"]}
 
         answers_clear.append(curr_answer_clear)
         answers_tokenized.append(curr_answer_tokenized)
     return answers_clear, answers_tokenized
 
+
+def create_samples_qa(dictionary, max_query_len, max_seq_len, doc_stride, n_special_tokens):
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/data_handler/samples.py:238]: We don't need it anymore - will delete when all tests pass

File: farm/data_handler/samples.py [hunk 7]
```diff
@@ -234,12 +226,85 @@ def process_answers(answers, doc_offsets, passage_start_c, passage_start_t):
...
+        # Deal with the potentially many answers (e.g. Squad or NQ dev set)
+        answers_clear, answers_tokenized = process_answers(dictionary["answers"],
+                                                           doc_offsets,
+                                                           passage_start_c,
+                                                           passage_start_t)
+
+        clear_text = {"passage_text": passage_text,
+                      "question_text": dictionary["question_text"],
+                      "passage_id": passage_id,
+                      "answers": answers_clear}
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/data_handler/samples.py:293]: This is not here because of a difference in format between Squad and NQ. I would like to neaten this section but was thinking of leaving it for our next clear up of the QA code

File: farm/data_handler/utils.py [hunk 8]
```diff
@@ -747,3 +748,65 @@ def filter_elements_per_worker(gen):
...
+        if follows_whitespace:
+            if ch not in string.whitespace:
+                map.append(i)
+                follows_whitespace = False
+        else:
+            if ch in string.whitespace:
+                follows_whitespace = True
+    return map
+
+def split_with_metadata(text):
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/data_handler/utils.py:767]: Ok the comment is in the latest commit

File: farm/infer.py [hunk 9]
```diff
@@ -293,16 +294,16 @@ def inference_from_file(self, file, multiprocessing_chunksize=None, streaming=Fa
...
     def inference_from_dicts(
-        self, dicts, rest_api_schema=False, multiprocessing_chunksize=None, streaming=False
+        self, dicts, return_json=True, num_processes=None, multiprocessing_chunksize=None, streaming=False
     ):
         """
         Runs down-stream inference on samples created from input dictionaries.
         The format of the input `dicts` depends on the task:
 
-        * QA: [{"qas": ["What is X?"], "context":  "Some context containing the answer"}]
+        * QA (internal):    [{"qas": ["What is X?"], "context":  "Some context containing the answer"}]
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/infer.py:302]: These two formats are so similar, I actually think it would be best if we just had the one input format. I think we should keep the Squad style dicts but call them something that is not specific to Squad

File: farm/infer.py [hunk 10]
```diff
@@ -335,8 +335,8 @@ def inference_from_dicts(
         if len(self.model.prediction_heads) > 0:
             aggregate_preds = hasattr(self.model.prediction_heads[0], "aggregate_preds")
 
-        if self.process_pool is None:  # multiprocessing disabled (helpful for debugging or using in web frameworks)
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/infer.py:338]: Yes, it's now fixed in the latest commit

File: farm/infer.py [hunk 11]
```diff
@@ -410,10 +412,10 @@ def _inference_with_multiprocessing(
         :rtype: iter
         """
 
-        # We group the input dicts into chunks and feed each chunk to a different process
-        # in the pool, where it gets converted to a pytorch dataset
-        results = self.process_pool.imap(
-            partial(self._create_datasets_chunkwise, processor=self.processor, rest_api_schema=rest_api_schema),
+        # We group the input dicts into chunks and feed each chunk to a different process,
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/infer.py]: Fixed in latest commit

File: farm/infer.py [hunk 12]
```diff
@@ -514,19 +524,20 @@ def _get_predictions_and_aggregate(self, dataset, tensor_names, baskets, rest_ap
...
         # In QA we can use this to get answers from long input texts by first getting predictions for smaller passages
         # and then aggregating them here.
 
         # can assume that we have only complete docs i.e. all the samples of one doc are in the current chunk
-        preds_all = self.model.formatted_preds(logits=[None], # For QA we collected preds per batch and do not want to pass logits
+        logits = [None] * len(unaggregated_preds_all)
+        preds_all = self.model.formatted_preds(logits=logits, # For QA we collected preds per batch and do not want to pass logits
                                                preds_p=unaggregated_preds_all,
-                                               baskets=baskets,
-                                               rest_api_schema=rest_api_schema)[0]
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/infer.py:529]: I suspect that is the case but I don't completely have my head around it yet. Also juggling the issue of having multiple prediction heads, multiple batches. I think I would like a proper clean up at some point where we implement classes which make clear exactly what kind of grouping of predictions we are looking at

File: farm/infer.py [hunk 13]
```diff
@@ -363,45 +363,47 @@ def inference_from_dicts(
...
+            preds_all = self._get_predictions(dataset, tensor_names, baskets)
+
+        if return_json:
+            preds_all = [x.to_json() for x in preds_all]
+
         return preds_all
 
     def _inference_with_multiprocessing(
-        self, dicts, rest_api_schema, aggregate_preds, multiprocessing_chunksize
+        self, dicts, return_json, aggregate_preds, multiprocessing_chunksize, num_processes
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/infer.py]: Fixed in latest commit

File: farm/utils.py [hunk 14]
```diff
@@ -391,5 +391,49 @@ def write_msmarco_results(results, output_filename):
...
+    start_ch = token_offsets[start_t]
+    # i.e. pointing at the END of the last token
+    if end_t == n_tokens:
+        end_ch = len(clear_text)
+    else:
+        end_ch = token_offsets[end_t]
+
+    return clear_text[start_ch: end_ch].strip(), start_ch, end_ch
 
+def pick_single_fn(heads, fn_name):
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/utils.py]: Done in latest commit

File: farm/modeling/adaptive_model.py [hunk 15]
```diff
@@ -358,18 +359,40 @@ def formatted_preds(self, logits, **kwargs):
...
+        preds_final = [list() for _ in range(n_heads)]
 
-        if len(self.prediction_heads) == 0:
+        if n_heads == 0:
             # just return LM output (e.g. useful for extracting embeddings at inference time)
-            all_preds = self.language_model.formatted_preds(logits=logits, **kwargs)
+            preds_final = self.language_model.formatted_preds(logits=logits, **kwargs)
+
+        elif n_heads == 1:
+            kwargs["preds_p"] = kwargs["preds_p"][0][0]
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/modeling/adaptive_model.py]: Ok will add a comment and check why it doesn't fail for other tasks

[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/modeling/adaptive_model.py]: In the latest commit, this is wrapped in a try catch to cover tasks other than QA. There is now a todo comment in the code saying that this should be tidied up

File: farm/modeling/prediction_head.py [hunk 16]
```diff
@@ -334,6 +333,9 @@ def forward(self, X):
 
     def logits_to_loss(self, logits, **kwargs):
         label_ids = kwargs.get(self.label_tensor_name)
+        # In Natural Questions, each dev sample can have multiple labels
+        # For loss calculation we only use the first label
+        label_ids = label_ids.narrow(1,0,1)
```
[the remark this answers is not in the document]
Comment (brandenchan) [reply farm/modeling/prediction_head.py:340]: The doc classification and ner examples still seem to run fine. narrow() should not have any effect on cases where there is only one label

**Comment by Timoeller** [comment]
I have thouroughly tested the code, especially changes to squad processing, but also other tasks like ner or doc_classification. This looks good.

2 Things NQ related: 
- The dataset in the examples/natural_questions.py cannot be loaded, please provide the download functionality
- When using the Inferencer on a NQ model (in tests/test_natural_questions.py) the output seems to be only squad style extractive QA. Is the classification head part integration missing?

**Comment by Timoeller** [comment]
There is still much missing but I think we could safely merge this branch without destroying too many other tasks : )

We have trained a model on NQ and evaluated on a NQ development set that has been converted into a format useable by haystack (it just comprises short answers or unanswerable questions).

I have hijacked SQuAD evaluation scripts to evaluate the NQ model and also a SQuAD model on this NQ development subset.
When just looking at the text predictions the NQ model performs much better.

-----------------------

NQ model:
"HasAns_exact": 43.16
"HasAns_f1": 52.04

SQuAD model:
"HasAns_exact": 25.0,
"HasAns_f1": 37.14
